src/Linters/PackagePrivateLinter.hack

```diff
@@ -0,0 +1,318 @@
+    foreach ($use_definition_context as $_key => $use_definition_context_value) {
+      foreach ($use_definition_context_value as $_inner_key => $use_definition_context_inner_value) {
+        if (!$this->isPrivateNamespacePathAllowed(
+          $use_definition_context_inner_value['parts'],
+          $current_namespace_for_file
+        )) {
+          return new ASTLintError(
+            $this,
+            'Package private artifacts are private to the parent package',
+            $use_definition_context_inner_value['node'],
+          );
+        }
+      }
+    }
+
+    foreach ($node->getDescendantsOfType(QualifiedName::class) as $qualified_name_token) {
+      if ($qualified_name_token->hasParts()) {
+        $qualified_name_parts = $this->getQualifiedNameFlatten(
+          $qualified_name_token,
+          $use_definition_context_shape['namespace'],
+          $current_namespace_for_file,
+        );
+        /**
+         * In some cases, the current namespace is also a qualified name,
+         * so it gets filtered into a qualified name token in our for each loop.
+         * We do not want to check for private keyword in that case.
+         */
+        if (C\count($qualified_name_parts) === C\count($current_namespace_for_file) &&
+          $qualified_name_parts === $current_namespace_for_file) {
+          continue;
+        }
+        if (!$this->isPrivateNamespacePathAllowed(
+          $qualified_name_parts,
+          $current_namespace_for_file,
+        )) {
+          return new ASTLintError(
+            $this,
+            'Package private artifacts can only be used within the top level parent package',
+            $qualified_name_token,
+          );
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * This method takes in a given qualified name and the current namespace for the file (could be empty).
+   * It checks if the qualified name has "private" in it, then it should be within the top level namespace.
+   * eg: \Test1\__Private could be used in either \Test1 or \Test1\Test2 namespaces.
+   */
+  private function isPrivateNamespacePathAllowed(
+    vec<string> $namepath_parts,
+    vec<string> $current_namespace_for_file,
+  ): bool {
+    $private_in_method_call = C\contains($namepath_parts, '_Private')
+      || C\contains($namepath_parts, '__Private');
+    if ($private_in_method_call) {
+      /**
+       * If current namespace is empty, do not allow to use private artifacts.
+       * Additionally, if the top level namespace does not match, we do not care for rest of the qualified path.
+       */
+      if (C\is_empty($current_namespace_for_file) || $current_namespace_for_file[0] !== $namepath_parts[0]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * This method returns the namespace for the file defined at the very top.
+   * eg: namespace TestNamespace; returns TestNamespace.
+   */
+  private function getTopLevelNamespace(this::TNode $node): ?string {
```

Comment:
mentioned above: think this is unneeded, you just need `$script->getNamespaces()`